responsibleai/responsibleai/_tools/shared/base_result.py

```diff
@@ -0,0 +1,149 @@
+# Copyright (c) Microsoft Corporation
+# Licensed under the MIT License.
+"""BaseResult of ResponsibleAI tools."""
+
+import json
+import jsonschema
+
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Any, Generic, List, TypeVar, Union
+
+from responsibleai.serialization_utilities import serialize_json_safe
+from responsibleai._tools.shared.attribute_serialization import (
+    load_attributes, save_attributes)
+from responsibleai._internal.constants import (
+    SerializationAttributes)
+
+
+TResult = TypeVar('TResult')  # Type for subclasses of BaseResult
+
+
+class BaseResult(ABC, Generic[TResult]):
+    def __init__(self, id: str, version: str) -> TResult:
+        """Construct a BaseResult.
+
+        :param id: ID of the result.
+        :param version: Version of the result.
+        :return: New BaseResult.
+        """
+        self.id = id
+        self.version = version
+
+        self._dashboard_data = None
+
+    def save(self, path: Union[str, Path]) -> None:
+        """Save the result to disk at the provided path.
+
+        :param path: Path to result directory on disk.
+        """
+        result_dir = Path(path)
+        result_dir.mkdir(parents=True, exist_ok=True)
+
+        # Save metadata like ID and and version to disk
+        self._save_metadata(result_dir)
+
+        # Save all result attributes to disk
+        save_attributes(self, self._get_attributes(), result_dir)
+
+        # Save dashboard data
+        self._save_dashboard(result_dir)
+
+    def _save_metadata(self, result_dir: Path):
+        """Save result metadata to disk.
+
+        :param path: Path to result directory on disk.
+        """
+        id_path = result_dir / SerializationAttributes.ID_FILENAME
```

Comment:
Just curious: why make the ID and version files separate? This isn't a particular objection for now, but in future, might they be combined in an `MLCausal` YAML file?

Reply:
This will have back-compat implications, so ideas for combining them are totally welcome. What else would the MLCausal YAML include @riedgar-ms?

Reply:
I've not thought that far yet. `MLCausal` `MLExplainer` etc. are still rather vague ideas

Reply:
I don't think it's worth holding up this PR for that. As you say, there are potential back-compat implications, but I don't think thoughts are sufficiently fixed to avoid them.

The `MLModel` file is shown on this page:
https://www.mlflow.org/docs/latest/models.html